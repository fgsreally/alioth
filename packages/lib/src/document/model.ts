import { interval } from '../interval'
import { VirtualDocument } from './document'
import type { NodeAttrs } from './node'

export class DocumentModel<T extends NodeAttrs> {
  activeId = '0'
  id = 0
  docs: { doc: VirtualDocument<T>; id: string; title: string }[] = []

  constructor() {
    interval.setState('$doc', this)
  }

  get activeDoc() {
    return this.find(this.activeId)!.doc!
  }

  get title() {
    return this.find(this.activeId)!.title
  }

  get container() {
    return this.activeDoc.root
  }

  get isActive() {
    return !!this.find(this.activeId)
  }

  get activeNode() {
    return this.find(this.activeId)!.doc?.activeNode
  }

  active(id: string) {
    if (this.activeId === id)
      return
    if (this.docs.some(item => item.id === id))
      this.activeId = id
  }

  find(id: string) {
    return this.docs.find(item => item.id === id)
  }

  load(data: any) {
    data = typeof data === 'string' ? JSON.parse(data) : data

    this.docs = data.map(({ id, title, data }: any) => {
      const doc = new VirtualDocument()
      doc.load(data)
      return {
        id, title, doc,

      }
    })
    return this.docs
  }
}
