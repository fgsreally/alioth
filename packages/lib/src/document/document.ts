import { debounce, omit } from 'lodash-es'
import { watch } from 'vue'
import type { Transaction, YEvent } from 'yjs'
import { Map as YMap } from 'yjs'
import { emitter } from 'phecda-vue'
import type { NodeAttrs } from './node'
import { VirtualNode } from './node'
import type { Controller } from './controller'
export class VirtualDocument<A extends NodeAttrs> {
  data: { root: VirtualNode<A>; activeNode?: VirtualNode<A>; hoverNode?: VirtualNode<A> }
  blockMap: Map<string, VirtualNode<A>> = new Map()
  controller: Controller
  root: VirtualNode<A>
  activeNode?: VirtualNode<A>
  hoverNode?: VirtualNode<A>
  emitter = emitter
  constructor(initAttrs?: A) {
    this.root = this.createNode(initAttrs, 'root')
  }

  get container() {
    return this.root
  }

  bindController(controller: Controller) {
    this.controller = controller
    this.root.bindDoc(this)
  }

  // 从所有block中找
  get(id: string) {
    return this.blockMap.get(id)
  }

  // 回收不在document中的block，垃圾回收
  clean() {
    this.blockMap.forEach((v, k, m) => {
      if (!v.parent)
        m.delete(k)
    })
  }

  transact(cb: () => void) {
    this.controller?.ydoc.transact(cb)
  }

  createNode(initAttrs?: A, id?: string) {
    const node = new VirtualNode<A>(initAttrs)

    id && (node.id = id)

    node.bindDoc(this)

    this.blockMap.set(node.id, node)

    return node
  }

  _createNode(initAttrs?: A, id?: string) {
    const node = new VirtualNode<A>(initAttrs)

    id && (node.id = id)

    node.doc = this

    this.blockMap.set(node.id, node)

    return node
  }

  removeNode(node: VirtualNode<NodeAttrs>) {
    if (node.parent) {
      node.parent.remove(node.index!)
      this.blockMap.delete(node.id)
    }
  }

  _removeNode(node: VirtualNode<NodeAttrs>) {
    if (node.parent) {
      node.parent._remove(node.index!)
      this.blockMap.delete(node.id)
    }
    else {
      this.blockMap.delete(node.id)
    }
  }

  select(node: VirtualNode<A>, type: 'activeNode' | 'hoverNode' = 'activeNode') {
    return this[type] = node
  }

  cancel(type: 'activeNode' | 'hoverNode' = 'activeNode') {
    return this[type] = undefined
  }

  load(data: any) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const that = this
    function traverse(data: any) {
      const node = new VirtualNode(data.key)

      node.bindDoc(that)

      node.id = data.id
      node.attrs = data.attrs
      data.children.forEach((item: any, i: number) => {
        const child = traverse(item)
        child.parent = node
        node.children[i] = child
      })
      return node
    }
    this.root = traverse(data)
    return this
  }

  bind(property: keyof VirtualNode<A>, timeout = 500) {
    const update = debounce((value: any) => {
      this.activeNode!.setAttribute(property, value)
    }, timeout)
    watch(() => this.activeNode!.attrs[property], (n: any) => update(n))
    return this.activeNode!.attrs[property]
  }
}

export function observe(doc: VirtualDocument<any>) {
  const fn: (arg0: YEvent<any>[], arg1: Transaction) => void = (events, t) => {
    events.forEach((event) => {
      if ((!t.local) || t.origin) { // from remote or undoManager
        if (event.changes.keys.size === 0) {
          event.changes.added.forEach((item: any) => {
            const id = item.content.getContent()[0]
            const node = doc.get(id)!
            if (item.left) {
              const left = doc.get(item.left.content.getContent()[0])!
              left.parent!._insert(node, left.index!)
            }
            else {
              const parent = doc.get(item.parent.parent.get('id'))
              parent!._insert(node, 0)
            }
          })
          // only work when undo

          event.changes.deleted.forEach((item) => {
            const id = item.content.getContent()[0]
            const node = doc.get(id)!
            node && node.parent?._remove(node.index!)
          })
        }
        else {
          event.changes.keys.forEach((item, i) => {
            const obj = event.target.get(i)

            switch (item.action) {
              case 'add':
                if (obj instanceof YMap)
                  doc._createNode(omit(event.target.get(i).toJSON(), ['children']), i)

                else doc.get(event.path[0] as string)?._setAttribute(i, obj)
                break
              case 'delete':
                if (item.oldValue instanceof YMap)
                  doc._removeNode(doc.get(i)!)
                else

                  doc.get(event.path[0] as string)?._setAttribute(i, obj)

                break
              case 'update':

                doc.get(event.path[0] as string)?._setAttribute(i, obj)
            }
          })
        }
      }
      else {
        // @ts-expect-error need phecda events type w
        doc?.emitter.emit('alioth:node-action', event)
      }
    })
  }
  doc.controller.map.observeDeep(fn)
  return () => {
    doc.controller.map.unobserveDeep(fn)
  }
}
