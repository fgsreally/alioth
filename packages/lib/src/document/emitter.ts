export class Emitter {
  private record: Record<string, Set<Function>> = {}
  on(event: string, cb: Function) {
    if (!this.record[event])
      this.record[event] = new Set()

    this.record[event].add(cb)
  }

  emit(event: string, data: any) {
    if (!this.record[event])
      return
    [...this.record[event]].forEach(cb => cb(data))
  }

  off(event: string, cb: Function) {
    this.record[event].delete(cb)
  }
}

export const emitter = new Emitter()
