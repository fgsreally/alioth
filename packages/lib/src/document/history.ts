// import { EventEmitter } from 'eventemitter3'
// import { ALIOTH_EVENT } from '../common'
// import type { NodeAttrs, VirtualNode } from './node'
// interface AppendEvent {
//   node: VirtualNode<any>
//   parent: VirtualNode<any>
//   type: 'append'
//   timestamp: number
// }

// interface RemoveEvent {
//   node: VirtualNode<any>
//   index: number
//   parent: VirtualNode<any>
//   type: 'remove'
//   timestamp: number
// }

// interface PropertyEvent {
//   type: 'property'
//   node: VirtualNode<any>
//   key: PropertyKey
//   value: any
//   timestamp: number

// }
// interface InsertEvent {
//   type: 'insert'
//   index: number
//   parent: VirtualNode<any>
//   timestamp: number

// }
//    type HistoryEvent = PropertyEvent | RemoveEvent | AppendEvent | InsertEvent

// export class HistoryController {
//   undoStack: HistoryEvent[][] = []
//   redoStack: HistoryEvent[][] = []
//   recordingEvents: HistoryEvent[]
//   isRecording = false
//   emitter = new EventEmitter()
//   private recordLayer: number

//   get isEndPoint() {
//     return this.redoStack.length === 0
//   }

//   recordStart(event: HistoryEvent | void) {
//     if (this.isRecording === false) {
//       if (event)
//         this.redoStack = []
//       this.recordingEvents = []
//       this.recordLayer = 0
//       this.isRecording = true
//     }
//     this.recordLayer++
//     event && this.recordingEvents.push(event)
//   }

//   recordEnd(category: 'undo' | 'redo') {
//     if (--this.recordLayer === 0) {
//       (category === 'undo' ? this.undoStack : this.redoStack).push(this.recordingEvents)
//       this.isRecording = false
//     }
//   }

//   undo() {
//     const events = this.undoStack.pop()
//     this.emitter.emit(ALIOTH_EVENT.UNDO)

//     if (events) {
//       this.recordStart()
//       // this.redoStack.push(events)
//       for (const ev of events)
//         this.switchAction(ev, 'redo')

//       this.recordEnd('redo')
//     }
//   }

//   redo() {
//     const events = this.redoStack.pop()
//     this.emitter.emit(ALIOTH_EVENT.REDO)

//     if (events) {
//       this.recordStart()
//       for (const ev of events)
//         this.switchAction(ev, 'undo')
//       this.recordEnd('undo')
//     }
//   }

//   switchAction(ev: HistoryEvent, category: 'redo' | 'undo') {
//     switch (ev.type) {
//       case 'remove':
//         ev.parent._insert(ev.node, ev.index, category)
//         break
//       case 'property':
//         ev.node._setAttribute(ev.key as unknown as keyof NodeAttrs, ev.value, category)
//         break
//       case 'insert' :
//         ev.parent._remove(ev.index, category)
//     }
//   }

//   clear() {
//     this.redoStack = []
//     this.undoStack = []
//   }
// }
