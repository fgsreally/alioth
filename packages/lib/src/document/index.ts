export * from './document'
// export * from './node'
export * from './controller'
