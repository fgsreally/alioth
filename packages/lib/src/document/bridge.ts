import type { VirtualDocument } from './document'
import { type Controller, type NodeEvent, applyEventToNode } from './controller'

export type CommitEvent = NodeEvent & { time: number }
export abstract class ClientBridge {
  protected readonly _memory = new Map<string, number>()// store event timestamp
  constructor(protected controller: Controller) {
    controller.invokeBridge = (event) => {
      const time = performance.now()
      if (event.type === 'set')

        this._memory.set(`set-${event.nodeId}-${event.key}`, time)

      if (event.type === 'swap')
        this._memory.set(`swap-${event.nodeId}`, time)
      this.send({ ...event, time })
    }
  }

  abstract send(event: CommitEvent): void

  handle(event: CommitEvent) {
    if (event.type === 'set') {
      const localUpdateTime = this._memory.get(`set-${event.nodeId}-${event.key}`)
      if (localUpdateTime && localUpdateTime > event.time)
        return
    }
    if (event.type === 'swap') {
      const localUpdateTime = this._memory.get(`swap-${event.nodeId}`)

      if (localUpdateTime && localUpdateTime > event.time)
        return
    }

    this.controller.applyEvent(event)
  }
}

export class ServerBridge {
  protected readonly _memory = new Map<string, number>()// store event timestamp
  constructor(protected doc: VirtualDocument) {

  }

  handle(event: CommitEvent) {
    if (event.type === 'set') {
      const localUpdateTime = this._memory.get(`set-${event.nodeId}-${event.key}`)
      if (localUpdateTime && localUpdateTime > event.time)
        return false

      this._memory.set(`set-${event.nodeId}-${event.key}`, event.time)
    }
    if (event.type === 'swap') {
      const localUpdateTime = this._memory.get(`swap-${event.nodeId}`)

      if (localUpdateTime && localUpdateTime > event.time)
        return false

      this._memory.set(`swap-${event.nodeId}`, event.time)
    }

    return applyEventToNode(this.doc, event)
  }
}

export class WsClientBridge extends ClientBridge {
  socket: WebSocket
  constructor(controller: Controller, protected url: string) {
    super(controller)
    this.socket = new WebSocket(url)
    this.socket.onmessage = (e) => {
      this.handle(e.data as any)
    }
  }

  send(e: NodeEvent) {
    this.socket.send(JSON.stringify(e))
  }
}
