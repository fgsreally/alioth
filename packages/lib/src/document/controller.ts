import { nanoid } from 'nanoid'
import { cloneDeep, isEqual } from 'lodash-es'
import EventEmitter from 'eventemitter3'
import { VirtualNode } from './document'
import type { VirtualDocument } from './document'

interface Options {
  length: number
}

export class Controller extends EventEmitter {
  options: Options
  undoStack: NodeEvent[] = []
  redoStack: NodeEvent[] = []

  currentEventId: string | undefined
  constructor(protected doc: VirtualDocument<any>, options: Partial<Options> = {}) {
    super()
    this.options = {
      length: 300,
      ...options,
    }

    doc.on('insert', ({ node }: any) => {
      this.redoStack = []
      this.addEvent({
        records: doc.flat(node).map(({ attrs, id, parent, index }) => {
          return {
            attrs,
            parentId: parent,
            index,
            nodeId: id,
          }
        }),
        type: 'insert',
        eventId: this.currentEventId || nanoid(),
      })
    })

    doc.on('swap', ({ node, lastParentId, lastIndex }: any) => {
      this.redoStack = []
      this.addEvent({
        parentId: node.parent,
        type: 'swap',
        lastParentId,
        index: node.index,
        lastIndex,
        nodeId: node.id,
        eventId: this.currentEventId || nanoid(),
      })
    })

    doc.on('remove', ({ node }: any) => {
      this.redoStack = []
      this.addEvent({
        type: 'remove',
        eventId: this.currentEventId || nanoid(),
        records: doc.flat(node).map(({ attrs, id, parent, index }) => {
          return {
            attrs,
            parentId: parent,
            index,
            nodeId: id,
          }
        }),
      })
    })

    doc.on('set', ({ node, key, value, oldValue }: any) => {
      this.redoStack = []
      this.addEvent({
        key,
        value: cloneDeep(value),
        type: 'set',
        oldValue: cloneDeep(oldValue),
        nodeId: node.id,
        eventId: this.currentEventId || nanoid(),
      })
    })
  }

  invokeBridge(_event: NodeEvent) {

  }

  transact(cb: () => void) {
    this.currentEventId = nanoid()
    cb()
    this.currentEventId = undefined
  }

  addEvent(event: NodeEvent, stack: 'undoStack' | 'redoStack' = 'undoStack') {
    if (this[stack].length >= this.options.length)
      this[stack].shift()

    this.invokeBridge(event)
    this[stack].push(event)
  }

  undo() {
    const event = this.undoStack.pop()

    if (event) {
      const { event: newEvent, isWork } = this.handleUndoEvent(event)

      this.addEvent(newEvent, 'redoStack')

      if (this.undoStack[this.undoStack.length - 1]?.eventId === event.eventId)
        this.undo()

      return isWork
    }

    return false
  }

  redo() {
    const event = this.redoStack.pop()
    if (event) {
      const { event: newEvent, isWork } = this.handleUndoEvent(event)

      this.addEvent(newEvent, 'undoStack')
      if (this.undoStack[this.redoStack.length - 1]?.eventId === event.eventId)
        this.redo()

      return isWork
    }
    return false
  }

  applyEvent(event: NodeEvent) {
    if (!applyEventToNode(this.doc, event))
      this.emit('error', event)
    else this.emit('success', event)
  }

  handleUndoEvent(event: NodeEvent) {
    switch (event.type) {
      case 'insert':
        event = {
          type: 'remove',
          eventId: event.eventId,
          records: event.records,
        }
        break

      case 'remove':
        event = {
          type: 'insert',
          eventId: event.eventId,
          records: event.records,
        }
        break
      case 'swap':
        event = {
          type: 'swap',
          eventId: event.eventId,
          index: event.lastIndex,
          lastIndex: event.index,
          lastParentId: event.parentId,
          parentId: event.lastParentId,
          nodeId: event.nodeId,
        }
        break
      case 'set':
        event = {
          type: 'set',
          eventId: event.eventId,
          nodeId: event.nodeId,
          key: event.key,
          value: event.oldValue,
          oldValue: event.value,

        }
    }

    return {
      isWork: this.applyEvent(event)!,
      event,
    }
  }
}

export function applyEventToNode(doc: VirtualDocument, event: NodeEvent) {
  if (event.type === 'insert') {
    const parentNode = doc.findById(event.records[0].parentId)

    if (!parentNode)
      return false

    event.records.forEach(({ attrs, nodeId, parentId, index }) => {
      const node = new VirtualNode(attrs, nodeId)
      node.parent = parentId
      node.index = index
      doc.nodeSet.add(node)
    })

    return true
  }

  if (event.type === 'remove') {
    const parentNode = doc.findById(event.records[0].parentId)

    if (!parentNode)
      return false
    const newNode = doc.findById(event.records[0].nodeId)
    if (!newNode)
      return false

    event.records.forEach(({ nodeId }) => {
      const node = doc.findById(nodeId)!
      node && doc.nodeSet.delete(node)
    })

    return true
  }
  if (event.type === 'swap') {
    const parentNode = doc.findById(event.parentId)!

    const newNode = doc.findById(event.nodeId)
    if (!newNode || !parentNode)
      return false

    newNode.parent = event.parentId
    newNode.index = event.index

    return true
  }

  if (event.type === 'set') {
    const newNode = doc.findById(event.nodeId)

    if (!newNode)
      return false

    const currentAttr = newNode.attrs[event.key]
    if (!isEqual(currentAttr, event.oldValue))
      event.oldValue = cloneDeep(currentAttr)

    newNode.attrs[event.key] = event.value

    return true
  }

  return false
}

interface BaseEvent {
  eventId: string
}

interface NodeRecord {
  nodeId: string
  parentId: string
  attrs: any
  index: number
}

export interface SwapEvent extends BaseEvent {
  type: 'swap'
  nodeId: string

  lastParentId: string
  lastIndex: number
  index: number
  parentId: string
}

export interface InsertEvent extends BaseEvent {
  type: 'insert'
  records: NodeRecord[]

}

export interface RemoveEvent extends BaseEvent {
  type: 'remove'

  records: NodeRecord[]
}

export interface SetEvent extends BaseEvent {
  type: 'set'
  nodeId: string

  key: string
  oldValue: any
  value: any
}

export type NodeEvent = SetEvent | InsertEvent | RemoveEvent | SwapEvent

// export function diff(nodes1: VirtualNode[], nodes2: VirtualNode[]): NodeEvent[] {
//   const events = [] as NodeEvent[]
//   const oldNodes1 = [] as VirtualNode[]
//   const oldNodes2 = [] as VirtualNode[]
//   if (node1.id !== node2.id)
//     throw new Error('node1 and node2 should have the same id to diff')

//   for (const n of node1.children) {
//     const node = node2.children.find(item => item.id === n.id)
//     if (!node) {
//       events.push({
//         ...traverseNode(n),
//         parentId: node1.id,
//         type: 'remove',
//         index: n.index,
//         eventId: nanoid(),
//       } as RemoveEvent)
//     }
//     else {
//       oldNodes1.push(n)
//     }
//   }

//   for (const n of node2.children) {
//     const node = node1.children.find(item => item.id === n.id)

//     if (!node) {
//       events.push({
//         ...traverseNode(n),
//         parentId: node1.id,
//         type: 'insert',
//         index: n.index,
//         eventId: nanoid(),
//       } as InsertEvent)
//     }
//     else {
//       oldNodes2.push(n)
//     }
//   }

//   oldNodes1.forEach((n, i) => {
//     const index = oldNodes2.findIndex(n2 => n2.id === n.id)!
//     if (index !== i) {
//       events.push({
//         type: 'swap',
//         from: i,
//         to: index,
//         nodeId: n.id,
//         eventId: nanoid(),
//         parentId: node1.id,
//       })
//     }
//     const n2 = oldNodes2[index]
//     for (const key in n.attrs) {
//       if (!isEqual(n.attrs[key], n2.attrs[key])) {
//         events.push({
//           nodeId: n.id,
//           eventId: nanoid(),
//           type: 'set',
//           key,
//           oldValue: n.attrs[key],
//           value: n2.attrs[key],
//         })
//       }
//     }
//     events.push(...diff(n, n2))
//   })

//   return events
// }

export function diff(nodes1: VirtualNode[], nodes2: VirtualNode[]) {
  const removeRecords = [] as NodeRecord[]
  const insertRecords = [] as NodeRecord[]
  const swapRecords = [] as {
    nodeId: string
    lastParentId: string
    lastIndex: number
    index: number
    parentId: string
  }[]

  const setRecords = [] as {
    nodeId: string
    key: string
    oldValue: any
    value: any
  }[]

  for (const n of nodes1) {
    const node = nodes2.find(node => node.id === n.id)
    if (!node) {
      removeRecords.push({
        nodeId: n.id,
        parentId: n.parent,
        attrs: n.attrs,
        index: n.index,
      })
    }
    else {
      if (node.index !== n.index || node.parent !== n.parent) {
        swapRecords.push({
          nodeId: n.id,
          lastParentId: n.parent,
          lastIndex: n.index,
          index: node.index,
          parentId: node.parent,
        })
      }

      for (const key in n.attrs) {
        if (!isEqual(n.attrs[key], node.attrs[key])) {
          setRecords.push({
            key,
            nodeId: n.id,
            oldValue: n.attrs[key],
            value: node.attrs[key],
          })
        }
      }
    }
  }
  for (const n of nodes2) {
    const node = nodes1.find(node => node.id === n.id)
    if (!node) {
      insertRecords.push({
        nodeId: n.id,
        parentId: n.parent,
        attrs: n.attrs,
        index: n.index,
      })
    }
  }
  return {
    insertRecords,
    removeRecords,
    swapRecords,
    setRecords,
  }
}

export function merge(base: VirtualNode[], branch1: VirtualNode[], branch2: VirtualNode[]) {
  const {
    insertRecords,
    removeRecords,
    swapRecords,
    setRecords,
  } = diff(base, branch2)
  const conflict = [] as any[]

  const newBranch = cloneDeep(branch1.filter(node => removeRecords.find(item => node.id === item.nodeId)))

  insertRecords.forEach((record) => {
    const { nodeId, parentId } = record
    const newNode = branch2.find(item => item.id === nodeId)!
    const parent = newBranch.find(item => item.id === parentId)

    if (parent)
      newBranch.push(newNode)
    else
      conflict.push(record)
  })
  setRecords.forEach((record) => {
    const { nodeId, key, value, oldValue } = record
    const node = newBranch.find(node => node.id === nodeId)!
    if (isEqual(oldValue, node.attrs[key]))
      node.attrs[key] = value

    else
      conflict.push(record)
  })

  swapRecords.forEach(({ nodeId, parentId, index }) => {
    const node = newBranch.find(node => node.id === nodeId)!

    //     const parent = newBranch.find(item => item.id === parentId)
    // if(!parent){
    //   conf
    // }
    node.parent = parentId
    node.index = index
  })

  branch2 = branch2.filter(node => removeRecords.find(item => node.id === item.nodeId))

  return { branch: branch1, conflict }
}
