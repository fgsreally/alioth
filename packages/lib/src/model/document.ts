import { nanoid } from 'nanoid'
import { Init } from 'phecda-vue'
import EventEmitter from 'eventemitter3'
import { VirtualDocument } from '../document'
import type { NodeAttrs } from '../document'
import { ALIOTH_EVENT } from '../common'
export class DocState<T extends NodeAttrs> extends EventEmitter {
  activeId: string
  containerAttrs: Record<string, any>
  docs: { doc: VirtualDocument<T>; id: string; title: string }[] = []

  @Init
  init() {
    // window.addEventListener('beforeunload', () => {
    //   localStorage.setItem('alioth_doc_state', JSON.stringify(this.docs.map((item) => {
    //     item.doc = item.doc.snapshot() as any
    //     return item
    //   })))
    // })
    const lastRecord = localStorage.getItem('alioth_doc_state') && false
    if (lastRecord) {
      this.docs = JSON.parse(lastRecord).map((item: any) => {
        const doc = new VirtualDocument()
        item.doc = doc.load(JSON.parse(item.doc))
        return item
      })
      this.active(this.docs[0]?.id || this.add())
    }

    else {
      this.active(this.add())
    }
  }

  get activeDoc() {
    return this.find(this.activeId)?.doc!
  }

  get title() {
    return this.find(this.activeId)!.title
  }

  get container() {
    return this.activeDoc.root
  }

  get isActive() {
    return !!this.activeNode
  }

  get activeNode() {
    return this.find(this.activeId)!.doc?.activeNode
  }

  add(title = '未定义') {
    const id = nanoid()
    const doc = new VirtualDocument('template', this.containerAttrs)
    this.docs.push({
      doc,
      id,
      title,
    })
    const emit = () => {
      this.emit('doc-action')
    }

    doc.HC.emitter.on(ALIOTH_EVENT.PROPERTY_CHANGE, emit)
    doc.HC.emitter.on(ALIOTH_EVENT.REMOVE_NODE, emit)
    doc.HC.emitter.on(ALIOTH_EVENT.APPEND_NODE, emit)

    return id
  }

  active(id: string) {
    if (this.activeId === id)
      return
    if (this.docs.some(item => item.id === id)) {
      // this.activeDoc.HC.emitter.removeAllListeners()
      this.activeId = id
    }
  }

  find(id: string) {
    return this.docs.find(item => item.id === id)
  }

  download(fileName: string) {
    const blob = new Blob([JSON.stringify(this.docs.map(({ doc, id, title }) => {
      return {
        id, title, data: doc.root,
      }
    }))])
    const saveLink = document.createElement('a')
    saveLink.href = URL.createObjectURL(blob)
    saveLink.download = fileName
    saveLink.click()
    URL.revokeObjectURL(saveLink.href)
  }

  remove(id: string) {
    if (this.docs.length > 1) {
      this.docs.splice(this.docs.findIndex(item => item.id === id), 1)[0].doc.unmount()
      if (!this.activeDoc)
        this.activeId = this.docs[0].id
    }
  }
}
