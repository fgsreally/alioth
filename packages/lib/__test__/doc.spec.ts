import { describe, expect, it } from 'vitest'
import { VirtualDocument, VirtualNode } from '../src/document/document'
describe('doc and node', () => {
  it('base action', () => {
    const doc = new VirtualDocument()

    const node1 = new VirtualNode({ id: 1 }, '1')
    const node2 = new VirtualNode({ id: 2 }, '2')
    doc.insert(node1, doc.root)
    doc.insert(node2, doc.root)

    expect(doc.nodes).toMatchSnapshot()

    doc.set(node1, 'id', 3)

    expect(node1.attrs.id).toBe(3)

    doc.insert(node2, node1)

    expect(node2.parent).toBe(node1.id)

    doc.remove(node1)

    expect(doc.nodes.length).toBe(0)
  })

  it('swap', () => {
    const doc = new VirtualDocument()

    const node1 = new VirtualNode({ id: 1 }, '1')
    const node2 = new VirtualNode({ id: 2 }, '2')
    doc.insert(node1, doc.root)
    doc.insert(node2, doc.root)
    expect(node1.index).toBe(0.5)
    expect(node2.index).toBe(0.25)
    doc.insert(node2, node1)

    expect(node2.index).toBe(0.5)
  })
})
