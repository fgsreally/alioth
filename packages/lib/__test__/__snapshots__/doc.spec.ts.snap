// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`doc and node > base action 1`] = `
[
  {
    "attrs": {
      "id": 1,
    },
    "id": "1",
    "index": 0.5,
    "parent": "root",
  },
  {
    "attrs": {
      "id": 2,
    },
    "id": "2",
    "index": 0.25,
    "parent": "root",
  },
]
`;
